packages/test-sync/src/commonTest/kotlin/io/realm/kotlin/test/mongodb/common/mongo/MongoCollectionTests.kt

```diff
@@ -0,0 +1,1466 @@
+                        ChildCollectionDataType::class
+                    )
+                )
+            )
+        )
+    }
+
+    @AfterTest
+    fun teadDown() {
+        app.asTestApp.run {
+            runBlocking {
+                COLLECTION_SCHEMAS.forEach {
+                    deleteDocuments(app.configuration.appId, it.simpleName!!, "{}")
+                }
+            }
+        }
+        if (this::app.isInitialized) {
+            app.close()
+        }
+    }
+
+    @Test
+    fun name() {
+        assertEquals("CollectionDataType", collection.name)
+    }
+
+    @Test
+    open fun withDocumentClass() = runBlocking<Unit> {
+        // Original typing
+        assertIs<Int>(collection.insertOne(CollectionDataType("object-1", Random.nextInt())))
+        assertIs<CollectionDataType>(collection.findOne())
+
+        // Reshaped
+        @OptIn(ExperimentalKBsonSerializerApi::class)
+        val bsonCollection: MongoCollection<BsonDocument> = collection.withDocumentClass()
+        assertIs<Int>(bsonCollection.insertOne(BsonDocument("_id" to BsonInt32(Random.nextInt()), "name" to BsonString("object-2"))))
+        assertIs<BsonDocument>(bsonCollection.findOne())
+        assertEquals(2, bsonCollection.count())
+    }
+
+    @Test
+    fun withDocumentClass_withCustomSerialization() = runBlocking<Unit> {
+        @OptIn(ExperimentalKBsonSerializerApi::class)
+        val reshapedCollectionWithDefaultSerializer: MongoCollection<CustomDataType> =
+            collection.withDocumentClass()
+
+        assertFailsWithMessage<SerializationException>("Serializer for class 'CustomDataType' is not found.") {
+            reshapedCollectionWithDefaultSerializer.insertOne(CustomDataType("object-2"))
+        }
+
+        @OptIn(ExperimentalKBsonSerializerApi::class)
+        val reshapedCollectionWithCustomSerializer: MongoCollection<CustomDataType> =
+            collection.withDocumentClass(customEjsonSerializer)
+
+        assertIs<Int>(reshapedCollectionWithCustomSerializer.insertOne(CustomDataType("object-2")))
+    }
+
+    @Test
+    fun count() = runBlocking<Unit> {
+        assertEquals(0, collection.count())
+
+        collection.insertMany((1..10).map { CollectionDataType("object-${it % 5}") })
+
+        assertEquals(10, collection.count())
+        assertEquals(5, collection.count(limit = 5))
+        assertEquals(2, collection.count(filter = BsonDocument("name" to BsonString("object-0"))))
+        assertEquals(2, collection.count(filter = BsonDocument("name", "object-0")))
+        assertEquals(
+            1,
+            collection.count(filter = BsonDocument("name" to BsonString("object-0")), limit = 1)
+        )
+    }
+
+    @Test
+    fun count_invalidFilter() = runBlocking<Unit> {
+        assertFailsWithMessage<ServiceException>("operator") {
+            collection.count(filter = BsonDocument("\$who", "object-0"))
+        }
+    }
+
+    @Test
+    open fun findOne() = runBlocking<Unit> {
```

Comment:
Why it has to be open?

Reply:
It was to overwrite the test in the actual implementations. I have remove the `open`. Actually it is now possible to select which implementing class you want to run the test for when you click the gutter icon for a test in an interface in latest Android Studio. That is quite neat.